resources/todomvc/architecture-examples/lit/src/lib/todos.ts

```diff
@@ -0,0 +1,182 @@
+the Software, and to permit persons to whom the Software is furnished to do so,
+subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE. */
+
+// This alphabet uses `A-Za-z0-9_-` symbols.
+// The order of characters is optimized for better gzip and brotli compression.
+// References to the same file (works both for gzip and brotli):
+// `'use`, `andom`, and `rict'`
+// References to the brotli default dictionary:
+// `-26T`, `1983`, `40px`, `75px`, `bush`, `jack`, `mind`, `very`, and `wolf`
+let urlAlphabet = "useandom-26T198340PX75pxJACKVERYMINDBUSHWOLF_GQZbfghjklqvwyzrict";
+
+function nanoid(size = 21) {
+    let id = "";
+    // A compact alternative for `for (var i = 0; i < step; i++)`.
+    let i = size;
+    while (i--) {
+        // `| 0` is more compact and faster than `Math.floor()`.
+        id += urlAlphabet[(Math.random() * 64) | 0];
+    }
+    return id;
+}
+
+export interface Todo {
+    id: string;
+    text: string;
+    completed: boolean;
+}
+
+export type TodoEdit = Partial<Todo> & { id: string };
+
+const todoFilters = ["all", "active", "completed"] as const;
+export type TodoFilter = (typeof todoFilters)[number];
+function isTodoFilter(value: string | undefined): value is TodoFilter {
+    return todoFilters.includes(value as TodoFilter);
+}
+
+/**
+ * A mutable, observable container for a todo list.
+ *
+ * @fires a `change` event when the todo list changes.
+ */
+export class Todos extends EventTarget {
+    #todos: Array<Todo> = [];
+    #filter: TodoFilter = this.#filterFromUrl();
+
+    get all(): ReadonlyArray<Todo> {
+        return this.#todos;
+    }
+
+    get active(): ReadonlyArray<Todo> {
+        return this.#todos.filter((todo) => !todo.completed);
+    }
+
+    get completed(): ReadonlyArray<Todo> {
+        return this.#todos.filter((todo) => todo.completed);
+    }
+
+    get allCompleted(): boolean {
+        return this.#todos.every((todo) => todo.completed);
+    }
+
+    connect() {
+        window.addEventListener("hashchange", this.#onHashChange);
+    }
+
+    disconnect() {
+        window.removeEventListener("hashchange", this.#onHashChange);
+    }
+
+    filtered() {
+        switch (this.#filter) {
+            case "active":
+                return this.active;
+            case "completed":
+                return this.completed;
+            case "all":
+            case undefined:
```

Comment:
we we need these two lines or is it always falling through and returning this.all?

Reply:
Didn't think about that! Done

Reply:
Good point! Done